src/Agent.Worker/Container/DockerCommandManager.cs

```diff
@@ -100,48 +101,82 @@ public async Task<int> DockerPull(IExecutionContext context, string image)
+                }
+                if (!string.IsNullOrEmpty(port.Protocol))
+                {
+                    portArg += $"/{port.Protocol}";
+                }
+                if (!string.IsNullOrEmpty(portArg))
+                {
+                    dockerOptions.Add(portArg);
                 }
             }
-
-            string node;
-            if (!string.IsNullOrEmpty(context.Container.ContainerBringNodePath))
+            dockerOptions.Add($"{container.ContainerCreateOptions}");
+            foreach (var env in container.ContainerEnvironmentVariables)
             {
-                node = context.Container.ContainerBringNodePath;
+                if (String.IsNullOrEmpty(env.Value) && String.IsNullOrEmpty(context.Variables.Get("_VSTS_DONT_RESOLVE_ENV_FROM_HOST")))
+                {
+                    // TODO: Remove fallback variable if stable
+                    dockerOptions.Add($"-e \"{env.Key}\"");
+                }
+                else
+                {
+                    dockerOptions.Add($"-e \"{env.Key}={env.Value.Replace("\"", "\\\"")}\"");
+                }
             }
-            else
+            foreach (var volume in container.MountVolumes)
             {
-                node = context.Container.TranslateToContainerPath(Path.Combine(HostContext.GetDirectory(WellKnownDirectory.Externals), "node", "bin", $"node{IOUtil.ExeExtension}"));
+                // replace `"` with `\"` and add `"{0}"` to all path.
+                String volumeArg;
+                if (String.IsNullOrEmpty(volume.SourceVolumePath))
+                {
+                    // Anonymous docker volume
+                    volumeArg = $"-v \"{volume.TargetVolumePath.Replace("\"", "\\\"")}\"";
+                }
+                else
+                {
+                    // Named Docker volume / host bind mount
+                    volumeArg = $"-v \"{volume.SourceVolumePath.Replace("\"", "\\\"")}\":\"{volume.TargetVolumePath.Replace("\"", "\\\"")}\"";
+                }
+                if (volume.ReadOnly)
+                {
+                    volumeArg += ":ro";
+                }
+                dockerOptions.Add(volumeArg);
             }
+            // IMAGE
+            dockerOptions.Add($"{container.ContainerImage}");
+            // COMMAND
+            dockerOptions.Add($"{container.ContainerCommand}");
+
+            var optionsString = string.Join(" ", dockerOptions);
+            List<string> outputStrings = await ExecuteDockerCommandAsync(context, "create", optionsString);
 
-            string sleepCommand = $"\"{node}\" -e \"setInterval(function(){{}}, 24 * 60 * 60 * 1000);\"";
-#if OS_WINDOWS
-            string dockerArgs = $"--name {displayName} --label {DockerInstanceLabel} {options} {dockerEnvArgs} {dockerMountVolumesArgs} {image} {sleepCommand}";  // add --network={network} and -v '\\.\pipe\docker_engine:\\.\pipe\docker_engine' when they are available (17.09)
```

Comment:
--network will be added even on Windows if the ContainerNetwork is set. --driver nat for networking works on server 1803, so I allow it, since we dont allow containers at all below this version

Reply:
do we need to update the hosted image to consume latest version docker for windows?

Reply:
I dont think so, at least not now. We do not allow `container:` on anything less than server 1803 anyways, which is our hosted windows container pool, so this will error out in the same way as it currently does for hosted vs2017, which is to say container feature is not allowed

For private agents the platform is obviously up to them, but if its windows I create with --driver nat which is the simplest option that will be supported by default on windows (and if they arent running > 1803 anyway, it will fail like before)

In general, nothing changes for our hosted windows, and for private windows agents they just need the server to support the nat network plugin which it almost always will